internal/manifest/manifest.go

```diff
@@ -0,0 +1,194 @@
+package manifest
+
+import (
+	"fmt"
+	"os"
+	"slices"
+
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/canonical/chisel/internal/jsonwall"
+	"github.com/canonical/chisel/internal/setup"
+)
+
+const schema = "1.0"
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name,omitempty"`
+	Version string `json:"version,omitempty"`
+	Digest  string `json:"sha256,omitempty"`
+	Arch    string `json:"arch,omitempty"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name,omitempty"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path,omitempty"`
+	Mode      string   `json:"mode,omitempty"`
+	Slices    []string `json:"slices,omitempty"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice,omitempty"`
+	Path  string `json:"path,omitempty"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+// Read loads a Manifest from a file without performing any validation. The file
+// is assumed to be both valid jsonwall and a valid Manifest (see [Validate]).
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	db, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+	mfestSchema := db.Schema()
+	if mfestSchema != schema {
+		return nil, fmt.Errorf("unknown schema version %q", mfestSchema)
+	}
+
+	manifest = &Manifest{db: db}
+	return manifest, nil
+}
+
+func (manifest *Manifest) IteratePaths(pathPrefix string, onMatch func(Path) error) (err error) {
+	return iteratePrefix(manifest, Path{Kind: "path", Path: pathPrefix}, onMatch)
+}
+
+func (manifest *Manifest) IteratePackages(onMatch func(Package) error) (err error) {
```

Comment:
We talked about using pointers for these but it's still using values. Is there a reason to keep these, or maybe I just misunderstood the plan?